src/emulator/auth/state.ts

```diff
@@ -171,10 +173,30 @@ export class ProjectState {
     } else {
       deleteProviders.push(PROVIDER_PHONE);
     }
+    if (user.mfaInfo) {
+      this.updateMfaEnrollments(user);
+    } else {
+      this.mfaEnrollmentIdsForLocalId.delete(user.localId);
+    }
 
     return this.updateUserProviderInfo(user, upsertProviders, deleteProviders);
   }
 
+  private updateMfaEnrollments(user: UserInfo): UserInfo {
+    for (const factor of user.mfaInfo ?? []) {
+      const allowUpdate = !!factor.mfaEnrollmentId;
+      const mfaEnrollmentId = factor.mfaEnrollmentId ?? randomId(28);
+      const enrollments = this.mfaEnrollmentIdsForLocalId.get(user.localId) ?? new Set();
+      if (!(enrollments.add(mfaEnrollmentId) || allowUpdate)) {
+        throw new Error("New MfaEnrollmentId wasn't unique.");
```

Comment:
This check also needs to happen in `operations.ts` with the right error code. This check here should only serve as an internal sanity check.

Reply:
@yuchenshi I am working on improving this validation. In doing so, it unearthed a few things worth considering.

There are three ways a user can be opted into MFA via the Firebase SDK. 

1. Using the Admin SDK `createUser` API 
2. Using the Admin SDK `updateUser` API. 
3.. Using the client SDK.

As far as I can tell, there is no support for the client SDK MFA enrollment flow in the Firebase Emulator (I could be wrong, here). So, we only need to focus on the Admin SDK entrypoints in the scope of this PR.

The schema for MFA Info on the sign up request doesn't specify an `mfaEnrollmentId` field, and moreoever, the Admin SDK [validates and rejects any attempt](https://github.com/firebase/firebase-admin-node/blob/master/src/auth/auth-api-request.ts#L1499-L1515) to specify an enrollment ID in the "createUser" flow. So, for the Admin SDK creation case, the MFA enrollment ID will always be [absent in the signUp flow](https://github.com/firebase/firebase-admin-node/blob/master/src/auth/user-import-builder.ts#L96-L113), and we will always need to [create a new enrollmentId](https://github.com/firebase/firebase-admin-node/blob/master/src/auth/auth-api-request.ts#L278).


Here's the underlying type for an instance of MFA Info on the create request, as well (no ID-like field present):

```
    GoogleCloudIdentitytoolkitV1MfaFactor: {
      /**
       * Display name for this mfa option e.g. "corp cell phone".
       */
      displayName?: string;
      /**
       * Phone number to receive OTP for MFA.
       */
      phoneInfo?: string;
    };
  ```
  
As I worked on this change, it seemed like this PR would be incomplete without extending support (and validation) to the API's exercised by the `updateUser` API as well. So, I am taking the spirit of this comment, and working it into the `signUp` and `setAccountInfo` flows, while also accounting for the fact that `mfaEnrollmentId` must not be specified on a `signUp` request. I am keeping these commits separate, so if there's a desire to limit the scope, we can evaluate rolling back some or all of the subsequent changes related to this.

Reply:
Great work so far. There's indeed no support for client-side MFA enrollment flow and yes, let's leave it like that.

As you noticed, `MfaFactor` only contains a subset fields (`displayName`, `phoneInfo`) compared to `MfaEnrollment`, making it impossible to specify an ID through `signUp`. Please kindly write a test to try to specify `mfaEnrollmentId` and see what happens. Best case, our routers will automatically block it. Otherwise, you may want to code defensively (e.g. picking only two fields manually from `reqBody.mfaInfo`). And I'd like to see that happen within this PR.

On the other hand,`setAccountInfo` seems to require `mfaEnrollmentId`, and that validation should be done in `operations.ts`. Can one reuse an existing second factor (but say, add one more)? What is the exact error code when adding new factor but `mfaEnrollmentId` isn't unique? And is uniqueness enforced for each account, or across all accounts? This can raise a lot of edge cases here. If you want to extend support, feel free to test against your real project and document the finding as tests. If you want to limit the scope, you can also choose to only support typical payloads issued by the admin SDK and throw `NotImplementedError` on anything else.

Reply:
I went ahead and extended support to the update case as well (along with a _hopefully_ spanning set of test cases). I really appreciate the thoughtful comment you left previously, here are some responses to the framing questions you posed.

> Can one reuse an existing second factor (but say, add one more)? 

Using the real backend, yes, a user can re-use an existing second factor, update an existing second factor, add another second factor, remove a second factor, remove all second factors. I attempted to support all of this behavior in the scope of this change.

> What is the exact error code when adding new factor but mfaEnrollmentId isn't unique?

  I created a distinct error for this scenario in the emulator (`INVALID_MFA_ID : enrollment ID was not unique.`). I’m not sure offhand how the admin API works against the real backend when trying to re-use an existing ID that is owned by another user, and if it throws an error, what the specific error is. I will make a note to test this locally.  
  
  > ... is uniqueness enforced for each account, or across all accounts?  
  
  My assumption — currently untested — is that uniqueness for the ID is enforced across all accounts. It’s worth noting, uniqueness is not enforced for the MFA phone number using the real API — any number of users can share the same underlying phone number for MFA.

There are a few behaviors resulting from how update is implemented as it pertains to “state” as managed by `src/emulators/auth/state.ts`. Certain updates may effectively “delete” a device corresponding to an “enrollment ID”, and in those scenarios the identifier for the deleted device will remain in the `localIdForMfaEnrollmentId` map, and similarly, the device is not removed from the set of ID’s maintained at `mfaEnrollmentIdsForLocalId`. These stateful maps are currently only cleared when `deleteAllAccounts` is called. It would be possible to extend this change to clean up the stateful maps storing MFA enrollment data, but it also seems like the current implementation may suffice for now. I think this only becomes an issue when a request to create a device has an ID collision with one of the deleted devices (which should be mitigated by randomness). Do you have an opinion here?

Reply:
> What is the exact error code when adding new factor but mfaEnrollmentId isn't unique?
tested this for you :) see comment inline

>   My assumption — currently untested — is that uniqueness for the ID is enforced across all accounts. 
This doesn't seem aligned with my quick glance at the code. Please make sure to double check this. It's very counterintuitive that one user will need to avoid duplicates with all other users and the error code may leak information.